multithreaded transaction admission

gandrewstone opened the pull request:
... by making the mempool read-only during parallel transaction processing, and then locking tx processing and committing the changes to the mempool.  Transactions can be processed in parallel because before parallel processing they are checked for conflicts with each other using a fastfilter (bloom-like filter) of the transaction inputs.  Any tx that conflicts with any tx that has been previously put on the parallel queue is deferred for evaluation after all current tx are committed.  Parallel threads threads therefore do not need to check for conflicts with other inflight transactions and so do not need to update the mempool right away.  Periodically another thread stops all tx admission threads and commits all accepted transactions to the mempool.

This is the major architectural change coming from the giga_perf branch.  Subsequent changes will be localized optimizations of the fastfilter, network, message processing, finer-grained locking, etc.

Code dealing with mempool admission (generally from main.cpp) has been isolated into its own files called
txadmission.cpp/h.

A new test mempool_accept.py is created that exercises the parallel acceptance code, and focuses on transaction conflicts (doublespends).

This PR creates a new rpc call "enqueuerawtransaction" that allows the caller to submit transactions very rapidly.  This call should replace "sendrawtransaction" for normal use.  However, when it returns it does not guarantee that the submitted transaction is in the mempool (unless "flush" is called) so "sendrawtransaction" is still useful in QA tests that may want to create a few tx and then mine them.
  
It also showcases the use of cashlib to generate and sign transactions for testing purposes.  With cashlib and enqueuerawtransaction, I have seen creation rates of >1500+ tx/sec on my 3ghz machine.

gandrewstone committed:
multithreaded transaction admission by making the mempool read-only during parallel transaction processing, and then locking tx processing and committing the changes to the mempool

ptschip commented:
Looks like you have a compile error

./../src/main.cpp:1741:25:   required from here
../../src/fastfilter.h:81:49: error: ‘bzero’ was not declared in this scope
     void reset() { bzero(&vData[0], FILTER_BYTES); }

sickpig commented on qa/rpc-tests/excessive.py:
```diff
@@ -655,9 +654,11 @@ def testExcessiveBlockSize(self):
                 try:
-                    self.nodes[random.randint(0, 3)].sendtoaddress(addrs[random.randint(0, 3)], .1)
+                    n = random.randint(0, 3)
+                    print(ntxs, "Send to %d" % n)
```
use logging framework

gandrewstone replied:
thx!

sickpig commented on qa/rpc-tests/excessive.py:
```diff
@@ -655,9 +654,11 @@ def testExcessiveBlockSize(self):
 
             addrs = [x.getnewaddress() for x in self.nodes]
             ntxs = 0
-            for i in range(0, random.randint(1, 200)):
```
why change it from 200 to 20?

gandrewstone replied:
trying to speed excessive.py up.  Its the slowest running test

sickpig commented on qa/rpc-tests/cashlibtest.py:
```diff
@@ -117,8 +115,7 @@ def run_test(self):
     bitcoinConf = {
-        "debug": ["net", "blk", "thin", "mempool", "req", "bench", "evict"],
-        "blockprioritysize": 2000000  # we don't want any transactions rejected due to insufficient fees...
+        "debug": ["rpc","net", "blk", "thin", "mempool", "req", "bench", "evict"],
```
nit: missing space between "rpc" and "net"

sickpig commented on qa/rpc-tests/cashlibtest.py:
```diff
@@ -31,7 +31,7 @@ def setup_network(self, split=False):
 
     def run_test(self):
-
+        FEE = 0
```
what's the purpose of this variable? Is used in the rest of the test to be subtracted to something else but since is set to 0...

guess that a comment to explain could be useful

gandrewstone replied:
It lets you create tx with different fees.  But a 0 fee tx is being accepted now so no need in the normal qa test to include a fee.

sickpig commented on qa/rpc-tests/excessive.py:
```diff
@@ -708,11 +709,43 @@ def info(type, value, tb):
     bitcoinConf = {
-        "debug": ["net", "blk", "thin", "mempool", "req", "bench", "evict"],  # "lck"
-        "blockprioritysize": 2000000  # we don't want any transactions rejected due to insufficient fees...
+        "debug": ["rpc","net", "blk", "thin", "mempool", "req", "bench", "evict"],
```
nit: space

sickpig commented on qa/rpc-tests/mempool_accept.py:
```diff
@@ -0,0 +1,474 @@
+
+        sighashtype = 0x41
+        n = 0
+        # print("amountin: %d amountout: %d outscript: %s" % (w["satoshi"], amt, w["scriptPubKey"]))
```
use log framework

sickpig commented on qa/rpc-tests/mempool_accept.py:
```diff
@@ -0,0 +1,474 @@
+    return count
+
+
+class MyTest (BitcoinTestFramework):
```
s/MyTest/MempoolAccept/

sickpig commented on qa/rpc-tests/mempool_accept.py:
```diff
@@ -0,0 +1,474 @@
+
+# Create a convenient function for an interactive python debugging session
+def Test():
+    t = MyTest(True)
```
s/MyTest/MempoolAccept/

sickpig commented on qa/rpc-tests/mempool_limit.py:
```diff
@@ -45,9 +45,45 @@ def run_test(self):
             txids.append([])
             txids[i] = create_lots_of_big_transactions(self.nodes[0], self.txouts, utxos[30*i:30*i+30], 30, (i+1)*base_fee)
 
+        print(self.nodes[0].getmempoolinfo())
```
use log framework

sickpig commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -130,15 +130,15 @@ def waitFor(timeout, fn, onError="timeout in waitFor", sleepAmt=1.0):
     while 1:
         result = fn()
-        if not result is None:
+        if not (result is None or result is False):
```
why this change?

gandrewstone replied:
Its a useability thing.  With this change I can do:
waitFor(10, lambda: node.something() == 50)

Without this change I have to explicitly return None like this:

waitFor(10, lambda: True if node.something() == 50 else None)

sickpig commented on src/Makefile.am:
```diff
@@ -246,6 +247,7 @@ libbitcoin_server_a_SOURCES = \
   iblt.cpp \
   init.cpp \
   dbwrapper.cpp \
+  rest.cpp \
```
why `rest.cpp` has been moved up rather than keep it in alphabetical order?

gandrewstone replied:
I was getting an undefined symbol but turned out link order was not the problem.  I will put it back

sickpig commented on src/chain.cpp:
```diff
@@ -17,9 +17,11 @@ void CChain::SetTip(CBlockIndex *pindex)
     if (pindex == NULL)
```
while at it check for `nullprt` rather than `NULL`

sickpig commented on src/fastfilter.h:
```diff
@@ -46,15 +46,15 @@ class CFastFilter
         vData.resize(FILTER_BYTES);
         // by sampling from a different part of the uint256, we make it harder for an attacker to generate collisions
-        grabFrom = (insecure_rand.rand32() % 7) * 4; // should be 4 byte aligned for speed
+        grabFrom = (1 + (insecure_rand.rand32() % 6)) * 4; // should be 4 byte aligned for speed
```
why this change?

gandrewstone replied:
the current fastfilter is a degenerate bloom filter.  It uses a k (numhashes) == 1.  This is a tradeoff between time and space since the optimum k is (expected entries)/(array size in bits) * ln2.  1 hash function is the fastest we can get.  But it would probably not be very much slower in the grand scheme of things  to offer k's up to 8 or 16 by breaking the 32byte hash into 4 byte chunks (and also the 4 byte chunks reversed).  But it'll likely be 8 or 16 times slower because we are doing 8 or 16 memory writes, which are the most expensive CPU instructions at this point.

Another advantage multiple k is that all the bits of the hash are actually used.  In the current fastfilter case only 32 bits are selected to be used.  If someone grinds these 32 bits to be all 0s we will get entirely conflicts.  Miners do this for block hashes so don't use those bits.  In review, I should probably xor all the bytes of the hash together into a single 32 byte chunk.  This would likely still be quite fast because it will pull from a single cache line.

Instead I grab one random chunk from 1 to 7 and then xor it (see below) with the location at 0.  This also guarantees to the user of fastfilter that location 0 is used in the hash.  This is useful if the user needs to stick additional information into the cryptographic hash since he can xor that additional info into the first byte and be certain that it will be part of the hash.  This happens in Bitcoin, especially in the tests, because what we really need to filter on is a unique COutput (a txid hash AND an integer index), not a unique txid.

sickpig commented on src/globals.cpp:
```diff
@@ -285,7 +312,8 @@ CTweakRef<bool> enableDataSigVerifyTweak("consensus.enableDataSigVerify",
 
-CTweak<unsigned int> numMsgHandlerThreads("net.msgHandlerThreads", "Number of message handler threads", 0);
+CTweak<unsigned int> numMsgHandlerThreads("net.msgHandlerThreads", "Max message handler threads", 0);
+CTweak<unsigned int> numTxAdmissionThreads("net.txAdmissionThreads", "Max transaction mempool admission threads", 0);
```
what if someone try to set a too high value here? couldn't we have a validato to limit the max number of thread to same sane value?

sickpig commented on src/main.cpp:
```diff
@@ -5336,8 +4838,10 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
-                bool fAlreadyHaveTx = AlreadyHaveTx(inv);
-                LOG(NET, "got inv: %s  %s peer=%d\n", inv.ToString(), fAlreadyHaveTx ? "have" : "new", pfrom->id);
+                bool fAlreadyHaveTx = TxAlreadyHave(inv);
+                // LOG(NET, "got inv: %s  %d peer=%s\n", inv.ToString(), fAlreadyHaveTx ? "have" : "new",
```
remove rather than comment?

sickpig commented on src/net.cpp:
```diff
@@ -2489,6 +2500,7 @@ void NetCleanup()
 
 void RelayTransaction(const CTransactionRef &ptx, const bool fRespend)
 {
+    // TODO call a GetSize() function that caches the tx size
```
:+1:

gandrewstone replied:
I have left a few optimization notes to ourselves in the code.

sickpig commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -165,12 +166,12 @@ BOOST_FIXTURE_TEST_CASE(uncache_coins, TestChain100Setup)
-    spends[2].vin[2].prevout.n = 0;
+    spends[2].vin[0].prevout.hash = coinbaseTxns[2].GetHash();
+    spends[2].vin[0].prevout.n = 0;
     spends[2].vout.resize(1);
```
don't get why we need to change `vin` order here

gandrewstone replied:
I wish that I could write a comment in the code to explain a change but then it automatically ages out or something.

In the mempool validation code, it used to check every input even if it found a problem in the first one.  I changed that to abort at the first missing input.  Because of this optimization, input 2 won't be loaded into the cache if input 0 was bad.

This test is checking proper cache loading for some inputs.  I need to put the inputs that the test is checking first, before bad ones.

sickpig commented on src/txadmission.cpp:
```diff
@@ -0,0 +1,1028 @@
+                empty = txInQ.empty() & txDeferQ.empty();
+            }
+            if (!empty)
+                MilliSleep(100);
```
why 100?

gandrewstone replied:
no real reason, could be 200 or 50.  I want to give enough time for other threads to work before wasting context switch time again, but don't want to be too slow.

sickpig commented on src/txadmission.cpp:
```diff
@@ -0,0 +1,1028 @@
+    }
+    if (!conflict)
+    {
+        //LOG(MEMPOOL, "Enqueue for processing %x\n", txd.tx->GetHash().ToString());
```
why comment this log statement since it is conditional to having `mempool` category enabled?

gandrewstone replied:
its extreme log spam.  Only useful if you are running a constrained test watching exactly what is going into the mempool, but in that case its very useful.  What we need is a telecom-style log level (like python logging.info, warning, error).  But without that, commenting out useful-but-spammy logs works.

sickpig commented on src/txadmission.cpp:
```diff
@@ -0,0 +1,1028 @@
+    }
+    else
+    {
+        //LOG(MEMPOOL, "Deferred %x\n", txd.tx->GetHash().ToString());
```
ditto

sickpig commented on src/wallet/wallet.cpp:
```diff
@@ -833,9 +843,9 @@ bool CWallet::AddToWallet(const CWalletTx &wtxIn, bool fFromLoadWallet, CWalletD
-        LOGA("AddToWallet %s  %s%s\n", wtxIn.GetHash().ToString(), (fInsertedNew ? "new" : ""),
-            (fUpdated ? "update" : ""));
+        // Only useful if debugging wallet
+        // LOGA("AddToWallet %s  %s%s\n", wtxIn.GetHash().ToString(), (fInsertedNew ? "new" : ""),
```
maybe we should add a new debug category for wallet related log statements

gandrewstone replied:
yes, but I didn't want to do it just for this one log.

sickpig commented:
@ptschip the compiler failure seem to be happening only when trying to cross-compile for win32/64 platform using mingw.

@gandrewstone I just give a quick review and left a bunch of minor feedback.

That said I think that before merging this we need to merge most of the ready PR on the queue, especially #1331 since it clear up the code for the upcoming Nov upgrade.

sickpig commented:
One another thing I noticed is the fact that to a bunch of py test we appended the definition of the `Test()` method. 

Two points about that: 

- I think we should find a place to better document how to use it 

- wouldn't be possible to add it to `util.py` (or somewhere else) so that we could avoid all that code repetition?

ptschip commented on src/chain.h:
```diff
@@ -429,11 +429,14 @@ class CChain
 private:
     std::vector<CBlockIndex *> vChain;
```
Won't compile on native windows
Missing:

#include <atomic>

ptschip commented on src/chain.h:
```diff
@@ -429,11 +429,14 @@ class CChain
 private:
     std::vector<CBlockIndex *> vChain;
```
oops , deleted that

Won't compile on native windows
Missing:

#include <atomic>

gandrewstone replied:
thanks

ptschip commented on src/chain.h:
```diff
@@ -429,11 +429,14 @@ class CChain
     CBlockIndex *Genesis() const { return vChain.size() > 0 ? vChain[0] : NULL; }
     /** Returns the index entry for the tip of this chain, or NULL if none. */
-    CBlockIndex *Tip() const { return vChain.size() > 0 ? vChain[vChain.size() - 1] : NULL; }
+    CBlockIndex *Tip() const { return tip; }
```
This does not return null anymore

-/** Returns the index entry for the tip of this chain, or NULL if none. */
+/** Returns the index entry for the tip of this chain */

gandrewstone replied:
tip can be nullptr in the same degenerate cases that Tip() returned null

ptschip commented on src/chain.h:
```diff
@@ -465,8 +468,8 @@ class CChain
-    int Height() const { return vChain.size() - 1; }
+    /** Return the maximal height in the chain. */
+    int Height() const { return tip.load() ? tip.load()->nHeight : -1; }
     /** Set/initialize a chain with a given tip. */
```
I take it we are now lock free here because tip is atomic...maybe a comment to that effect?

+/** Return the maximal height in the chain.  Does not require cs_main  */

gandrewstone replied:
ok, also added the same note for Tip()

ptschip commented on src/globals.cpp:
```diff
@@ -209,6 +210,32 @@ CNodeSignals g_signals;
+CFastFilter<4 * 1024 * 1024> incomingConflicts GUARDED_BY(csTxInQ);
+
+// Tranactions that are waiting for validation and are known not to conflict with others
+std::queue<CTxInputData> txInQ GUARDED_BY(csTxInQ);
```
Would be useful to have a short write up of the work flow of how a transaction arrives and gets processed using the different Q's.

gandrewstone replied:
I have a nice writeup IIRC.  I'll dig it up.

ptschip replied:
I think the write up is the last thing we need

ptschip commented on src/txadmission.cpp:
```diff
@@ -0,0 +1,1028 @@
+        entry.UpdateRuntimeSigOps(resourceTracker.GetSigOps(), resourceTracker.GetSighashBytes());
+
+#if 0 // to be removed -- bugs in this should have been shaken out
+        // Check again against just the consensus-critical mandatory script
```
I argue this code really should stay in. IMO We're asking for trouble by not checking for Mandatory script flags.  I also think this will break xval.

gandrewstone replied:
that's fine, we can argue about this local optimization later :-)

ptschip commented on src/main.cpp:
```diff
@@ -6874,10 +6225,20 @@ bool SendMessages(CNode *pto)
+            std::vector<uint256> blockHashesToAnnounce;
+            {
+                TRY_LOCK(pto->cs_inventory, locked);
+                if (locked)
```
I think when we have blocks to announce we should take the lock here, deferring is only going to slow down block prop.  In addition I have testing evidence that using try locks in SendMessages is a contributing factor to the burstiness of network traffic.

ptschip commented on src/main.cpp:
```diff
@@ -6733,19 +6096,6 @@ bool SendMessages(CNode *pto)
-        }
-        TRY_LOCK(pto->cs_vSend, lockSend);
-        if (!lockSend)
-        {
```
excellent..i've wanted to take these out for a while now and came to the same conclusion that they were a contributer to the burstiness of network traffic.

ptschip commented on src/main.cpp:
```diff
@@ -6886,19 +6247,25 @@ bool SendMessages(CNode *pto)
-                        continue;
-                    CBlockIndex *pindex = mi->second;
-                    if (chainActive[pindex->nHeight] != pindex)
+                    CBlockIndex *pindex = nullptr;
```
Looks good, but just a thought for future optimization, we could just as well store the header rather than the hash when we have a block to announce and forgo the need for all the subsequent lookups and cs_main locks.

gandrewstone replied:
good idea

ptschip commented on src/net.cpp:
```diff
@@ -384,6 +384,17 @@ CNodeRef FindNodeRef(const std::string &addrName)
 }
 
+CNodeRef FindNodeRef(const NodeId id)
+{
```
this function already exists in connmgr.cpp

gandrewstone replied:
thanks!  I didn't look around because FindNode() is here.  I've made a note to myself to move FindNode over into connmgr in a different "cleanup" PR.

ptschip committed:
Fix transactions per second

We should update txn per second for every txn that enters the mempool.
Due to the giga-net refactoring it was accidentally misplace.

ptschip commented on src/wallet/rpcwallet.cpp:
```diff
@@ -426,8 +426,6 @@ UniValue sendtoaddress(const UniValue &params, bool fHelp)
 
-    LOCK2(cs_main, pwalletMain->cs_wallet);
-
     CTxDestination dest = DecodeDestination(params[0].get_str());
```
are you sure about taking these locks out...seems we would at least need cs_wallet?

gandrewstone replied:
The locks are moved.
The code is read-only in the wallet until it gets to CWallet::CommitTransaction.  There I need to not have cs_main locked while doing AcceptToMemoryPool because that function now needs to let the tx admission processing threads run.  After AcceptToMemoryPool, the cs_main and cs_wallet locks are taken.

What I think may happen in the worst case is if you call sendtoaddress simultaneously from different processes, the 2 CreateTransaction calls may end up creating 2 tx that use the same inputs.  One of these will be rejected during AcceptToMemoryPool so one of the two sendtoaddress calls will fail with an exception.  However, I do not think that anyone is using the wallet with the kind of load required to trigger this exception since it can't handle that kind of load anyway.

ptschip commented on src/txadmission.cpp:
```diff
@@ -0,0 +1,1028 @@
+            {
+                for (const COutPoint &remove : vCoinsToUncache)
+                    pcoinsTip->Uncache(remove);
+                if (fMissingInputs)
```
-for (const COutPoint &remove : vCoinsToUncache)
-pcoinsTip->Uncache(remove);

We keep coins from orphans in the cache...if the orphan is later timed out of the orphan cache we remove them at that time. Reason is because the orphan will in most cases be pulled into the mempool and so we don't have to go and get the coins again, possibly from disk.

gandrewstone replied:
you're right, this code should be in an else clause of "if (fMissingInputs)" so that we only uncache if the transaction is rejected for another reason.

ptschip commented on src/txadmission.cpp:
- boost::unique_lock<boost::mutex> lock(csCommitQ);
+ std::unique_lock<std::mutex> lock(csCommitQ);

gandrewstone replied:
all of sync.h is using boost::mutex and boost::unique_locks, which means that almost all of our code is using the boost variants.  I think that in a different PR we should move these over all together.

ptschip commented on src/txadmission.cpp:
```diff
@@ -0,0 +1,1028 @@
+
+            // move the previously deferred txs into active processing
+            if (!txDeferQ.empty())
+            {
```
txDeferQ needs a lock

gandrewstone replied:
done

ptschip commented on src/txadmission.cpp:
```diff
@@ -0,0 +1,1028 @@
+{
+    std::vector<uint256> whatChanged;
+    LOCK(cs_main); // cs_main must lock before csCommitQ
+    {
```
why a lock on cs_main, isnt  it txCommitQ that needs a lock...or do we possibly need both here ?  LOCK2(cs_main , cstxCommitQ); ?

ptschip replied:
oh , duh I didn't see the lock on csCommitQ, I was expecting to see WRITELOCK...ok that looks good

ptschip commented on src/txadmission.h:
```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2018 The Bitcoin Unlimited developers
```
need to add txadmission.h and txadmission.cpp to  .formatted-files

gandrewstone replied:
done

ptschip commented on src/txadmission.cpp:
```diff
@@ -356,6 +355,12 @@ void ThreadTxAdmission()
+                {
+                    for (const COutPoint &remove : vCoinsToUncache)
+                        pcoinsTip->Uncache(remove);
+                }
```
this is actually already handled on L412 txadmission.cpp. (in accepttomemorypool)

gandrewstone replied:
AcceptToMemoryPool is now a top-level API which accepts a tx in the old serialized way.  It will (for example) return whether the tx was accepted or not.  

ThreadTxAdmission is the parallel version.  So the code needs to be in both places.  In a future PR, I want to change up AcceptToMemoryPool to use the parallel code, but wait for the result of the particular tx it is interested in.  This will involve putting extra data in CTxInputData and CTxCommitData so once the tx is processed it knows to notify some other thread.

gandrewstone committed:
format the new files

gandrewstone committed:
Merge pull request #25 from ptschip/thezerg_txnpersec

Fix transactions per second

gandrewstone commented:
the win32 test problem has been reproduced on dev and does not involve code changed by this PR.  Merging now.
